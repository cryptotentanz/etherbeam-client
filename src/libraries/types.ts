import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum UserType {
  User = 'user',
  Administrator = 'administrator',
}

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface User {
  name: string
  email: string
  type: UserType
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string | null
  price?: BigNumber | null
  priceHistory: TokenPriceHistory[]
  website?: string | null
  whitepaper?: string | null
  github?: string | null
  linkedin?: string | null
  facebook?: string | null
  reddit?: string | null
  twitter?: string | null
  telegram?: string | null
  discord?: string | null
  actions?: TransactionAction[] | null
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number | null
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address | null
  to?: Address | null
  value?: BigNumber | null
  gas: {
    unitPrice?: BigNumber | null
    limit?: number | null
    used?: number | null
    ratio?: number | null
    price?: BigNumber | null
  }
  action?: TransactionMethod | null
  logs?: TransactionMethod[] | null
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[] | null

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction?: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token | null
  fromAmount?: BigNumber | null
  toAddress?: Address | Token | null
  toAmount?: BigNumber | null
}
