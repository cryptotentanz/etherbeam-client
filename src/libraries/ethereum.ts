import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
