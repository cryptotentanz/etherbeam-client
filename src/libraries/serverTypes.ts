import {
  AddressType,
  TransactionActionType,
  TransactionMethodParameterType,
  TransactionStatus,
  UserType,
} from './types'

export interface UserData {
  email: string
  name: string
  user_type: UserType
}

export interface AddressData {
  address_hash: string
  address_type: AddressType
  label: string
}

export interface TransactionMethodParameterData {
  name: string
  parameter_type: TransactionMethodParameterType
  raw_type: string
  raw_value?: string
  decimal_value?: string
  addresses?: {
    address_hash: string
    address?: AddressData
  }[]
}

export interface TokenData extends AddressData {
  name: string
  symbol: string
  decimals: number
  chart_pair?: string
  description?: string
  price?: string
  price_1h?: string
  price_1h_ratio?: number
  price_1d?: string
  price_1d_ratio?: number
  price_1w?: string
  price_1w_ratio?: number
  price_1m?: string
  price_1m_ratio?: number
  price_1y?: string
  price_1y_ratio?: number
  website?: string
  whitepaper?: string
  github?: string
  linkedin?: string
  facebook?: string
  reddit?: string
  twitter?: string
  telegram?: string
  discord?: string
  from_transaction_actions?: TokenActionData[]
  to_transaction_actions?: TokenActionData[]
}

export interface TransactionData {
  transaction_hash: string
  status: TransactionStatus
  datetime: Date
  block_number?: number
  from_address_hash?: string
  from_address?: AddressData
  to_address_hash?: string
  to_address?: AddressData
  value?: string
  gas_unit_price?: string
  gas_limit?: number
  gas_used?: number
  gas_ratio?: number
  gas_fee?: string
  transaction_method_action?: TransactionMethodData
  transaction_method_logs?: TransactionMethodData[]
}

export interface TransactionMethodData {
  contract: AddressData
  name: string
  parameters: TransactionMethodParameterData[]
}

export interface TokenActionData {
  block_transaction: TransactionData
  index: number
  action_type: TransactionActionType
  holder_address_hash: string
  holder_address?: AddressData
  from_address_hash?: string
  from_address?: AddressData | TokenData
  from_amount?: string
  to_address_hash?: string
  to_address?: AddressData | TokenData
  to_amount?: string
}
