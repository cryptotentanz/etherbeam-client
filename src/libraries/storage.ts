import Cookies, { CookieSetOptions } from 'universal-cookie'
import { AuthToken } from '@totentech/api-client'

const cookies = new Cookies()

export const removeAuthToken = (): void => {
  const options: CookieSetOptions = {
    path: '/',
  }
  cookies.remove('tokenType', options)
  cookies.remove('client', options)
  cookies.remove('uid', options)
  cookies.remove('accessToken', options)
  cookies.remove('expiry', options)
}

export const saveAuthToken = (authToken: AuthToken | null): void => {
  if (!authToken) return removeAuthToken()

  const { client, uid, accessToken, tokenType, expiry } = authToken
  const options: CookieSetOptions = {
    path: '/',
    expires: new Date(parseInt(expiry) * 1000),
    // TODO: Add the secure option when HTTPS
    // secure: true,
    sameSite: 'strict',
  }

  cookies.set('tokenType', tokenType, options)
  cookies.set('client', client, options)
  cookies.set('uid', uid, options)
  cookies.set('accessToken', accessToken, options)
  cookies.set('expiry', expiry, options)
}

export const loadAuthToken = (): AuthToken | null => {
  const authToken = {
    tokenType: cookies.get('tokenType'),
    client: cookies.get('client'),
    uid: cookies.get('uid'),
    accessToken: cookies.get('accessToken'),
    expiry: cookies.get('expiry'),
  }
  return !Object.values(authToken).filter((v) => !v).length ? authToken : null
}
