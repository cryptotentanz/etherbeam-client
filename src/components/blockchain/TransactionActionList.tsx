import React, { FC, useState } from 'react'
import { Pagination } from '@totentech/api-client'
import { TransactionAction } from 'libraries/types'
import Table from '../Table'
import PaginatedList from '../PaginatedList'
import TransactionActionRow from './TransactionActionRow'

const ACTIONS_PER_PAGE = 20

interface TransactionActionsListProps {
  actions: TransactionAction[]
}

const TransactionActionsList: FC<TransactionActionsListProps> = ({ actions }: TransactionActionsListProps) => {
  const [page, setPage] = useState<number>(1)
  const pagination: Pagination = {
    currentPage: page,
    itemsPerPage: ACTIONS_PER_PAGE,
    totalItems: actions.length,
    totalPages: Math.ceil(actions.length / ACTIONS_PER_PAGE),
  }
  const { currentPage, itemsPerPage } = pagination
  const currentPageActions = actions.slice((currentPage - 1) * itemsPerPage, currentPage * itemsPerPage)

  return (
    <PaginatedList pagination={pagination} setPage={setPage}>
      <Table>
        <thead>
          <tr>
            <th>Action</th>
            <th>Time</th>
            <th className="my-d-min-xs">Holder</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {currentPageActions.map((action, index) => (
            <TransactionActionRow key={index} action={action} />
          ))}
        </tbody>
      </Table>
    </PaginatedList>
  )
}

export default TransactionActionsList
