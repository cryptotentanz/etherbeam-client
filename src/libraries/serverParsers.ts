import { BigNumber } from 'ethers'
import orderBy from 'lodash/orderBy'
import {
  TransactionData,
  TokenData,
  AddressData,
  TokenActionData,
  TransactionMethodParameterData,
  TransactionMethodData,
  UserData,
} from './serverTypes'
import {
  Transaction,
  Token,
  HistoryTime,
  Address,
  AddressType,
  TransactionMethodValueType,
  TransactionAction,
  TransactionActionDirection,
  TransactionMethodParameter,
  TransactionMethodParameterType,
  TransactionMethod,
  User,
} from './types'
import { createTransactionAddress, getEtherscanUrl } from './ethereum'

const parseBigNumberFromServer = (value: string): BigNumber => BigNumber.from(value.split('.')[0])

export const parseUserData = ({ name, email, user_type }: UserData): User => {
  return {
    name,
    email,
    type: user_type,
  }
}

export const parseAddressData = ({ address_hash, address_type, label }: AddressData): Address => {
  return {
    hash: address_hash,
    type: address_type,
    label,
    url: getEtherscanUrl('address', address_hash),
  }
}

const parseHashData = (hash: string, address?: AddressData | TokenData): Address => {
  if (!address) {
    return {
      hash,
      type: AddressType.Unknown,
      url: getEtherscanUrl('address', hash),
    }
  }

  switch (address.address_type) {
    case AddressType.Token:
      return parseTokenData(address as TokenData)
    default:
      return parseAddressData(address)
  }
}

const parseTransactionMethodValueType = ({
  parameter_type,
  decimal_value,
  addresses,
}: TransactionMethodParameterData): TransactionMethodValueType => {
  switch (parameter_type) {
    case TransactionMethodParameterType.UnsignedInteger:
      return decimal_value ? parseBigNumberFromServer(decimal_value) : null
    case TransactionMethodParameterType.Address:
      return addresses && addresses?.length > 0 ? parseHashData(addresses[0].address_hash, addresses[0].address) : null
    case TransactionMethodParameterType.Addresses:
      return addresses ? addresses.map((address) => parseHashData(address.address_hash, address.address)) : null
    default:
      return null
  }
}

const parseTransactionMethodParameter = (parameter: TransactionMethodParameterData): TransactionMethodParameter => {
  const { name, parameter_type, raw_type } = parameter

  return {
    name,
    type: parameter_type,
    rawType: raw_type,
    value: parseTransactionMethodValueType(parameter),
  }
}

const parseTransactionMethodData = ({ contract, name, parameters }: TransactionMethodData): TransactionMethod => {
  return {
    contract: parseAddressData(contract),
    name,
    parameters: parameters.map((parameter) => parseTransactionMethodParameter(parameter)),
  }
}

export const parseTransactionData = ({
  transaction_hash,
  status,
  datetime,
  block_number,
  from_address_hash,
  from_address,
  to_address_hash,
  to_address,
  value,
  gas_unit_price,
  gas_limit,
  gas_used,
  gas_ratio,
  gas_fee,
  transaction_method_action,
  transaction_method_logs,
}: TransactionData): Transaction => {
  return {
    blockNumber: block_number,
    dateTime: datetime,
    address: createTransactionAddress(transaction_hash),
    status,
    from: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    to: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    value: value ? parseBigNumberFromServer(value) : null,
    gas: {
      unitPrice: gas_unit_price ? parseBigNumberFromServer(gas_unit_price) : null,
      limit: gas_limit,
      used: gas_used,
      ratio: gas_ratio,
      price: gas_fee ? parseBigNumberFromServer(gas_fee) : null,
    },
    action: transaction_method_action ? parseTransactionMethodData(transaction_method_action) : null,
    logs: transaction_method_logs?.map((log) => parseTransactionMethodData(log)),
  }
}

const parseTokenActionData = (
  {
    block_transaction,
    holder_address_hash,
    holder_address,
    index,
    action_type,
    from_address,
    from_address_hash,
    from_amount,
    to_address,
    to_address_hash,
    to_amount,
  }: TokenActionData,
  direction: TransactionActionDirection
): TransactionAction => {
  return {
    transaction: parseTransactionData(block_transaction),
    type: action_type,
    direction,
    index,
    holder: parseHashData(holder_address_hash, holder_address),
    fromAddress: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    fromAmount: from_amount ? parseBigNumberFromServer(from_amount) : null,
    toAddress: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    toAmount: to_amount ? parseBigNumberFromServer(to_amount) : null,
  }
}

export const parseTokenData = (data: TokenData): Token => {
  const {
    address_hash,
    name,
    description,
    symbol,
    decimals,
    chart_pair,
    price,
    price_1h,
    price_1h_ratio,
    price_1d,
    price_1d_ratio,
    price_1w,
    price_1w_ratio,
    price_1m,
    price_1m_ratio,
    price_1y,
    price_1y_ratio,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    from_transaction_actions,
    to_transaction_actions,
  } = data
  const priceHistory = []
  if (price_1h && price_1h_ratio) {
    priceHistory.push({
      time: HistoryTime.Hour,
      price: parseBigNumberFromServer(price_1h),
      ratio: price_1h_ratio as number,
    })
  }
  if (price_1d && price_1d_ratio) {
    priceHistory.push({
      time: HistoryTime.Day,
      price: parseBigNumberFromServer(price_1d),
      ratio: price_1d_ratio as number,
    })
  }
  if (price_1w && price_1w_ratio) {
    priceHistory.push({
      time: HistoryTime.Week,
      price: parseBigNumberFromServer(price_1w),
      ratio: price_1w_ratio as number,
    })
  }
  if (price_1m && price_1m_ratio) {
    priceHistory.push({
      time: HistoryTime.Month,
      price: parseBigNumberFromServer(price_1m),
      ratio: price_1m_ratio as number,
    })
  }
  if (price_1y && price_1y_ratio) {
    priceHistory.push({
      time: HistoryTime.Year,
      price: parseBigNumberFromServer(price_1y),
      ratio: price_1y_ratio as number,
    })
  }

  const fromActions = from_transaction_actions
    ? from_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Sell))
    : []
  const toActions = to_transaction_actions
    ? to_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Buy))
    : []
  const actions =
    from_transaction_actions || to_transaction_actions
      ? orderBy([...fromActions, ...toActions], ['transaction.dateTime', 'index'], ['desc', 'desc'])
      : null

  return {
    ...parseAddressData(data),
    type: AddressType.Token,
    url: getEtherscanUrl('token', address_hash),
    name,
    description,
    symbol,
    decimals,
    chartPair: chart_pair,
    price: price ? parseBigNumberFromServer(price) : null,
    priceHistory,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    actions,
  }
}
