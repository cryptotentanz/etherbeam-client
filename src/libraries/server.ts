import compact from 'lodash/compact'
import {
  HttpStatus,
  AuthToken,
  RestAction,
  callApi,
  ApiClient,
  createApiClient,
  setAuthToken,
  subscribeToNewTokens,
  flushAuthToken,
} from '@totentech/api-client'
import { Token, USDC_HASH, User, WETH_HASH } from './types'
import { TokenData, UserData } from './serverTypes'
import { parseTokenData, parseUserData } from './serverParsers'
import { loadAuthToken, removeAuthToken, saveAuthToken } from './storage'

const TIMEOUT = 60 * 1000

type UserUpdateCallback = (user: User | null) => void

export interface ServerData {
  apiClient: ApiClient
  userSubscriptions: UserUpdateCallback[]
}

const onNewAuthToken = (data: ServerData, token: AuthToken): void => {
  saveAuthToken(token)

  if (!token) {
    onUserUpdate(data, null)
  }
}

const initializeAuthToken = (apiClient: ApiClient): void => {
  const token = loadAuthToken()

  if (token) setAuthToken(apiClient, token)
}

export const initializeServerData = (): ServerData => {
  const onUserUpdateCallback = (user: User | null): void => {
    if (!user) removeAuthToken()
  }
  const data = {
    apiClient: createApiClient(process.env.SERVER_API_HOST as string, true, TIMEOUT),
    userSubscriptions: [onUserUpdateCallback],
  }

  initializeAuthToken(data.apiClient)
  subscribeToNewTokens(data.apiClient, (token) => onNewAuthToken(data, token))

  return data
}

export const subscribeToUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  userSubscriptions.push(callback)
}

export const unsubscribeFromUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  const index = userSubscriptions.indexOf(callback)

  if (index < 0) return

  userSubscriptions.splice(index, 1)
}

const onUserUpdate = ({ userSubscriptions }: ServerData, user: User | null): void => {
  userSubscriptions.forEach((callback) => callback(user))
}

export const signIn = async (serverData: ServerData, email: string, password: string): Promise<User | null> => {
  const payload = { email, password }

  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Post,
    'auth/sign_in',
    [HttpStatus.Ok],
    null,
    payload,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const validateToken = async (serverData: ServerData): Promise<User | null> => {
  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Get,
    'auth/validate_token',
    [HttpStatus.Ok],
    null,
    null,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const signOut = async (data: ServerData): Promise<void> => {
  const { apiClient } = data
  await callApi(apiClient, RestAction.Delete, 'auth/sign_out', [HttpStatus.Ok, HttpStatus.NotFound], null, null, true)

  flushAuthToken(apiClient)
  onUserUpdate(data, null)
}

export const fetchTokens = async (serverData: ServerData): Promise<Token[]> => {
  const { data } = await callApi<TokenData[]>(serverData.apiClient, RestAction.Get, 'contract_tokens')

  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (
  serverData: ServerData,
  address: string,
  withList?: 'actions'
): Promise<Token | null> => {
  const { status, data } = await callApi<TokenData>(
    serverData.apiClient,
    RestAction.Get,
    `contract_tokens/${address}?list=${withList || ''}`,
    [HttpStatus.Ok, HttpStatus.NotFound]
  )

  if (status != HttpStatus.Ok) return null

  return parseTokenData(data)
}
