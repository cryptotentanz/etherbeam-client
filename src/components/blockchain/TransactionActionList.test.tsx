import * as React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { createTokenAction } from 'tests/fixtures/ethereum'
import { TransactionAction, TransactionActionType } from 'libraries/types'
import TransactionActionList from './TransactionActionList'

describe('<TransactionActionList />', () => {
  let component: ShallowWrapper

  describe('Default', () => {
    const actions = [
      createTokenAction(TransactionActionType.Transfer),
      createTokenAction(TransactionActionType.Transfer),
    ]

    beforeAll(() => {
      component = shallow(<TransactionActionList actions={actions} />)
    })

    it('should render rows', () => {
      const rows = component.find('TransactionActionRow')
      expect(rows.length).toBe(2)
      expect(rows.at(0).prop('action')).toBe(actions[0])
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Empty', () => {
    const actions: TransactionAction[] = []

    beforeAll(() => {
      component = shallow(<TransactionActionList actions={actions} />)
    })

    it('should not render rows', () => expect(component.find('TransactionActionRow').length).toBe(0))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })
})
