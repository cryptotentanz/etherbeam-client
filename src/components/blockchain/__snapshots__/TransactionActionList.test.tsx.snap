// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionActionList /> Default should match snapshot 1`] = `
<table
  class="my-table my-dark-bg"
>
  <thead>
    <tr>
      <th>
        Action
      </th>
      <th>
        Time
      </th>
      <th
        class="my-d-min-xs"
      >
        Holder
      </th>
      <th />
    </tr>
  </thead>
  <tbody>
    <tr
      class="my-dark-bg my-transactionactionrow my-dark-bg"
    >
      <td>
        <div
          class="my-transactionactionrow-action"
        >
          <span
            class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-alt-circle-right fa-w-16 "
                data-icon="arrow-alt-circle-right"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zM140 300h116v70.9c0 10.7 13 16.1 20.5 8.5l114.3-114.9c4.7-4.7 4.7-12.2 0-16.9l-114.3-115c-7.6-7.6-20.5-2.2-20.5 8.5V212H140c-6.6 0-12 5.4-12 12v64c0 6.6 5.4 12 12 12z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
          <span>
            <span>
              <span
                class="my-badge my-light-bg  "
              >
                <span
                  class="my-tokenamount-symbol"
                >
                  TKN-1
                </span>
                <span
                  class="my-tokenamount-amount"
                >
                   NaN
                </span>
              </span>
              <span>
                 to 
              </span>
              <span>
                <a
                  class="my-clickable  my-link my-secondary-fg"
                  href="https://etherscan.io/address/0x0000000000000000000000000000000000000222"
                  rel="noreferrer"
                  tabindex="0"
                  target="_blank"
                >
                  <span
                    class="my-badge my-light-bg  my-clickable"
                  >
                    <span
                      class="my-icon  "
                    >
                      <span
                        class=" "
                      >
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-coins fa-w-16 "
                          data-icon="coins"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 512 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
                            fill="currentColor"
                          />
                        </svg>
                      </span>
                      <span
                        class="my-icon-label "
                      >
                        Token 2 (TKN-2)
                      </span>
                    </span>
                  </span>
                </a>
              </span>
            </span>
          </span>
        </div>
      </td>
      <td
        class="my-text-center"
      >
        <span
          class="my-time"
        >
          11:22
        </span>
      </td>
      <td
        class="my-d-min-xs my-text-center"
      >
        <span>
          <a
            class="my-clickable  my-link my-light-fg"
            href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="my-badge my-light-bg  my-clickable"
            >
              <span
                class="my-icon  "
              >
                <span
                  class=" "
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-wallet fa-w-16 "
                    data-icon="wallet"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span
                  class="my-icon-label "
                >
                  My Wallet
                </span>
              </span>
            </span>
          </a>
        </span>
      </td>
      <td
        class="my-text-center"
      >
        <a
          class="my-clickable my-size-l my-ulink my-secondary-fg"
          href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
          rel="noreferrer"
          tabindex="0"
          target="_blank"
          title="Etherscan"
        >
          <div
            class="my-etherscanlink-img"
          >
            <img
              src="/images/etherscan.svg"
            />
          </div>
        </a>
      </td>
    </tr>
    <tr
      class="my-dark-bg my-transactionactionrow my-dark-bg"
    >
      <td>
        <div
          class="my-transactionactionrow-action"
        >
          <span
            class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-alt-circle-right fa-w-16 "
                data-icon="arrow-alt-circle-right"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zM140 300h116v70.9c0 10.7 13 16.1 20.5 8.5l114.3-114.9c4.7-4.7 4.7-12.2 0-16.9l-114.3-115c-7.6-7.6-20.5-2.2-20.5 8.5V212H140c-6.6 0-12 5.4-12 12v64c0 6.6 5.4 12 12 12z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
          <span>
            <span>
              <span
                class="my-badge my-light-bg  "
              >
                <span
                  class="my-tokenamount-symbol"
                >
                  TKN-1
                </span>
                <span
                  class="my-tokenamount-amount"
                >
                   NaN
                </span>
              </span>
              <span>
                 to 
              </span>
              <span>
                <a
                  class="my-clickable  my-link my-secondary-fg"
                  href="https://etherscan.io/address/0x0000000000000000000000000000000000000222"
                  rel="noreferrer"
                  tabindex="0"
                  target="_blank"
                >
                  <span
                    class="my-badge my-light-bg  my-clickable"
                  >
                    <span
                      class="my-icon  "
                    >
                      <span
                        class=" "
                      >
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-coins fa-w-16 "
                          data-icon="coins"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 512 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
                            fill="currentColor"
                          />
                        </svg>
                      </span>
                      <span
                        class="my-icon-label "
                      >
                        Token 2 (TKN-2)
                      </span>
                    </span>
                  </span>
                </a>
              </span>
            </span>
          </span>
        </div>
      </td>
      <td
        class="my-text-center"
      >
        <span
          class="my-time"
        >
          11:22
        </span>
      </td>
      <td
        class="my-d-min-xs my-text-center"
      >
        <span>
          <a
            class="my-clickable  my-link my-light-fg"
            href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="my-badge my-light-bg  my-clickable"
            >
              <span
                class="my-icon  "
              >
                <span
                  class=" "
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-wallet fa-w-16 "
                    data-icon="wallet"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span
                  class="my-icon-label "
                >
                  My Wallet
                </span>
              </span>
            </span>
          </a>
        </span>
      </td>
      <td
        class="my-text-center"
      >
        <a
          class="my-clickable my-size-l my-ulink my-secondary-fg"
          href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
          rel="noreferrer"
          tabindex="0"
          target="_blank"
          title="Etherscan"
        >
          <div
            class="my-etherscanlink-img"
          >
            <img
              src="/images/etherscan.svg"
            />
          </div>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<TransactionActionList /> Empty should match snapshot 1`] = `
<table
  class="my-table my-dark-bg"
>
  <thead>
    <tr>
      <th>
        Action
      </th>
      <th>
        Time
      </th>
      <th
        class="my-d-min-xs"
      >
        Holder
      </th>
      <th />
    </tr>
  </thead>
  <tbody />
</table>
`;
